src/mlpack/tests/rbm_network_test.cpp

```diff
@@ -0,0 +1,172 @@
+  BinaryLayer<> hidden(hiddenLayerSize, trainData.n_rows, 0);
+  GaussianInitialization gaussian(0, 0.1);
+  RBM<GaussianInitialization, BinaryLayer<>, BinaryLayer<>> model(trainData,
+      gaussian, visible, hidden, 1,  true);
+  model.Reset();
+  // Set the parmaeters from a learned rbm sklearn random state 23
+  model.Parameters() = arma::mat(
+      "-0.23224054, -0.23000632, -0.25701271, -0.25122418, -0.20716651,"
+      "-0.20962217, -0.59922456, -0.60003836, -0.6, -0.625, -0.475;");
+  // Check Weight Shared
+  BOOST_REQUIRE_CLOSE(arma::accu(model.VisibleLayer().Weight() -
+      model.HiddenLayer().Weight()), 0, 1e-14);
+
+  // Check free energy
+  arma::vec freeEnergy = arma::mat(
+      "-0.87523715, 0.50615066, 0.46923476, 1.21509084;");
+  arma::vec calcultedFreeEnergy(4);
+  calcultedFreeEnergy.zeros();
+  for (size_t i = 0; i < trainData.n_cols; i++)
+  {
+    model.VisibleLayer().ForwardPreActivation(std::move(trainData.col(i)),
+        std::move(output));
+    calcultedFreeEnergy(i) = model.FreeEnergy(std::move(trainData.col(i)));
+  }
+  for (size_t i = 0; i < freeEnergy.n_elem; i++)
+    BOOST_REQUIRE_CLOSE(calcultedFreeEnergy(i), freeEnergy(i), 1e-5);
+}
+BOOST_AUTO_TEST_CASE(MiscTest)
+{
+  /**
+   * Train and evaluate a vanilla network with the specified structure.
+   */
+
+  arma::mat X = arma::mat("0, 0, 0;"
+                          "0, 1, 1;"
+                          "1, 0, 1;"
+                          "1, 1, 1;");
+  std::cout << X.n_rows << std::endl;
+  std::cout << X.n_cols << std::endl;
+  X = X.t();
+  BuildVanillaNetwork<>(X, 2);
+}
+
+BOOST_AUTO_TEST_CASE(ClassificationTest)
+{
+  // Normalised dataset
+  int hiddenLayerSize = 100;
+  arma::mat trainData, testData, dataset;
+  arma::vec trainLabelsTemp, testLabelsTemp;
+  data::Load("mnist_data_small.csv", dataset, true);
+  trainData = dataset.submat(0, 0,
+      dataset.n_rows - 102, dataset.n_cols - 2);
+  testData = dataset.submat(dataset.n_rows - 102, 0 ,
+      dataset.n_rows - 2, dataset.n_cols - 2);
+  trainLabelsTemp = dataset.submat(0, dataset.n_cols - 1,
+      dataset.n_rows - 102, dataset.n_cols - 1);
+  testLabelsTemp = dataset.submat(dataset.n_rows - 102, dataset.n_cols - 1,
+      dataset.n_rows - 2, dataset.n_cols - 1);
+
+  arma::Row<size_t> trainLabels = arma::zeros<arma::Row<size_t>>(1,
+      trainLabelsTemp.n_rows);
+  arma::Row<size_t> testLabels = arma::zeros<arma::Row<size_t>>(1,
+      testLabelsTemp.n_rows);
+
+  for (size_t i = 0; i < trainLabelsTemp.n_rows; ++i)
+    trainLabels(i) = arma::as_scalar(trainLabelsTemp.row(i));
+
+  for (size_t i = 0; i < testLabelsTemp.n_rows; ++i)
+    testLabels(i) = arma::as_scalar(testLabelsTemp.row(i));
+
+  std::cout << arma::size(trainLabels) << std::endl;
+  std::cout << arma::size(testLabels) << std::endl;
+
+  trainData = trainData.t();
+  testData = testData.t();
+
+  arma::mat output, XRbm(hiddenLayerSize, trainData.n_cols),
+      YRbm(hiddenLayerSize, trainLabels.n_cols);
```

Comment:
I think I found the fix. You don't define the dimensions of `YRbm` properly.
Try the following:
`YRbm(hiddenLayerSize, testData.n_cols)`